diff --git a/src/main/java/org/springframework/data/mapping/model/ReflectionEntityInstantiator.java b/src/main/java/org/springframework/data/mapping/model/ReflectionEntityInstantiator.java
index 9d4ec5fb..a0ab2107 100644
--- a/src/main/java/org/springframework/data/mapping/model/ReflectionEntityInstantiator.java
+++ b/src/main/java/org/springframework/data/mapping/model/ReflectionEntityInstantiator.java
@@ -131,7 +131,7 @@ enum ReflectionEntityInstantiator implements EntityInstantiator {
 					ctype = ctype.getComponentType();
 					dims++;
 				}
-				return (T) Array.newInstance(clazz, dims);
+				return (T) Array.newInstance(ctype, new int[dims]);
 			} else {
 				return BeanUtils.instantiateClass(entity.getType());
 			}
diff --git a/src/test/java/org/springframework/data/mapping/model/ClassGeneratingEntityInstantiatorUnitTests.java b/src/test/java/org/springframework/data/mapping/model/ClassGeneratingEntityInstantiatorUnitTests.java
index 00bedb42..b6a44f2c 100755
--- a/src/test/java/org/springframework/data/mapping/model/ClassGeneratingEntityInstantiatorUnitTests.java
+++ b/src/test/java/org/springframework/data/mapping/model/ClassGeneratingEntityInstantiatorUnitTests.java
@@ -75,6 +75,46 @@ class ClassGeneratingEntityInstantiatorUnitTests<P extends PersistentProperty<P>
 		this.instance.createInstance(entity, provider);
 	}
 
+	@Test
+	void instantiatesOneDimensionalObjectArrayWithCorrectRuntimeType() {
+
+		doReturn(String[].class).when(entity).getType();
+		Object result = this.instance.createInstance(entity, provider);
+		assertThat(result.getClass()).isEqualTo(String[].class);
+	}
+
+	@Test
+	void instantiatesOneDimensionalPrimitiveArrayWithCorrectRuntimeType() {
+
+		doReturn(int[].class).when(entity).getType();
+		Object result = this.instance.createInstance(entity, provider);
+		assertThat(result.getClass()).isEqualTo(int[].class);
+	}
+
+	@Test
+	void instantiatesTwoDimensionalObjectArrayWithCorrectRuntimeType() {
+
+		doReturn(String[][].class).when(entity).getType();
+		Object result = this.instance.createInstance(entity, provider);
+		assertThat(result.getClass()).isEqualTo(String[][].class);
+	}
+
+	@Test
+	void instantiatesThreeDimensionalObjectArrayWithCorrectRuntimeType() {
+
+		doReturn(String[][][].class).when(entity).getType();
+		Object result = this.instance.createInstance(entity, provider);
+		assertThat(result.getClass()).isEqualTo(String[][][].class);
+	}
+
+	@Test
+	void instantiatesTwoDimensionalPrimitiveArrayWithCorrectRuntimeType() {
+
+		doReturn(int[][].class).when(entity).getType();
+		Object result = this.instance.createInstance(entity, provider);
+		assertThat(result.getClass()).isEqualTo(int[][].class);
+	}
+
 	@Test // DATACMNS-1126
 	void instantiatesTypeWithPreferredConstructorUsingParameterValueProvider() {
 
diff --git a/src/test/java/org/springframework/data/mapping/model/ReflectionEntityInstantiatorUnitTests.java b/src/test/java/org/springframework/data/mapping/model/ReflectionEntityInstantiatorUnitTests.java
index 16ca8ea5..0ec39209 100755
--- a/src/test/java/org/springframework/data/mapping/model/ReflectionEntityInstantiatorUnitTests.java
+++ b/src/test/java/org/springframework/data/mapping/model/ReflectionEntityInstantiatorUnitTests.java
@@ -63,6 +63,46 @@ class ReflectionEntityInstantiatorUnitTests<P extends PersistentProperty<P>> {
 		INSTANCE.createInstance(entity, provider);
 	}
 
+	@Test
+	void instantiatesOneDimensionalObjectArrayWithCorrectRuntimeType() {
+
+		doReturn(String[].class).when(entity).getType();
+		Object result = INSTANCE.createInstance(entity, provider);
+		assertThat(result.getClass()).isEqualTo(String[].class);
+	}
+
+	@Test
+	void instantiatesOneDimensionalPrimitiveArrayWithCorrectRuntimeType() {
+
+		doReturn(int[].class).when(entity).getType();
+		Object result = INSTANCE.createInstance(entity, provider);
+		assertThat(result.getClass()).isEqualTo(int[].class);
+	}
+
+	@Test
+	void instantiatesTwoDimensionalObjectArrayWithCorrectRuntimeType() {
+
+		doReturn(String[][].class).when(entity).getType();
+		Object result = INSTANCE.createInstance(entity, provider);
+		assertThat(result.getClass()).isEqualTo(String[][].class);
+	}
+
+	@Test
+	void instantiatesThreeDimensionalObjectArrayWithCorrectRuntimeType() {
+
+		doReturn(String[][][].class).when(entity).getType();
+		Object result = INSTANCE.createInstance(entity, provider);
+		assertThat(result.getClass()).isEqualTo(String[][][].class);
+	}
+
+	@Test
+	void instantiatesTwoDimensionalPrimitiveArrayWithCorrectRuntimeType() {
+
+		doReturn(int[][].class).when(entity).getType();
+		Object result = INSTANCE.createInstance(entity, provider);
+		assertThat(result.getClass()).isEqualTo(int[][].class);
+	}
+
 	@Test // DATACMNS-1126
 	void instantiatesTypeWithPreferredConstructorUsingParameterValueProvider() {
 
